### PR TITLE
Obtain match output elements without materializing the output.

### LINKED ISSUE #267: `Regex.Match` element accessors should not materialize the whole output
Currently `Regex.Match` element accessors are implemented this way:
```swift
  /// Lookup a capture by name or number
  public subscript<T>(dynamicMember keyPath: KeyPath<Output, T>) -> T {
    output[keyPath: keyPath]
  }

  // Allows `.0` when `Match` is not a tuple.
  @_disfavoredOverload
  public subscript(
    dynamicMember keyPath: KeyPath<(Output, _doNotUse: ()), Output>
  ) -> Output {
    output
  }
```

This is not correct as we should not materialize the entire output.

### PR DESCRIPTION
**Pull request by rxwei**
Resolves #267.

### CODE REVIEW

**Comment by rxwei** [comment]
@swift-ci please test

**Comment by rxwei** [comment]
@swift-ci please test

**Comment by rxwei** [comment]
@swift-ci please test

File: Sources/_RegexParser/Utility/TypeConstruction.swift [hunk 1]
```diff
@@ -139,3 +139,45 @@ public enum TypeConstruction {
...
+    guard let byteOffset = offset(of: keyPath) else {
+      return nil
+    }
+    if byteOffset == 0 { return 0 }
+    var currentOffset = 0
+    for (index, type) in elementTypes.enumerated() {
+      func sizeAndAlignMask<T>(_: T.Type) -> (Int, Int) {
+        (MemoryLayout<T>.size, MemoryLayout<T>.alignment - 1)
+      }
+      // The ABI of an offset-based key path only stores the byte offset, so
```
Comment (rxwei) [inline Sources/_RegexParser/Utility/TypeConstruction.swift:165]: @lorentey just reminded me of this... 😭
```console
> \(Void, Void).0 == \(Void, Void).1
$R0: Bool = true
```

@Azoy, do you have the radar number for this issue I can put in the comment?

**Comment by rxwei** [comment]
@swift-ci please test

File: Sources/_RegexParser/Utility/TypeConstruction.swift (hunk 1 above)
Comment (Azoy) [reply Sources/_RegexParser/Utility/TypeConstruction.swift:165]: rdar://63819465

File: Sources/_StringProcessing/Regex/Match.swift [hunk 2]
```diff
@@ -37,33 +41,48 @@ extension Regex.Match {
...
+  var wholeMatchType: Any.Type {
+    value.map { type(of: $0) } ?? Substring.self
+  }
+
   /// Accesses a capture by its name or number.
   public subscript<T>(dynamicMember keyPath: KeyPath<Output, T>) -> T {
-    output[keyPath: keyPath]
+    // Note: We should be able to get the element offset from the key path
+    // itself even at compile time. We need a better way of doing this.
+    guard let outputTupleOffset = MemoryLayout.tupleElementIndex(
```
Comment (rxwei) [inline Sources/_StringProcessing/Regex/Match.swift:76]: When we store the whole match inside the capture list, we will be able to precompute these offsets for all elements in a `Regex.Match`. I'll do that next.

**Comment by rxwei** [comment]
@swift-ci please test

File: Sources/_StringProcessing/Regex/AnyRegexOutput.swift [hunk 3]
```diff
@@ -62,9 +62,11 @@ extension AnyRegexOutput: RandomAccessCollection {
...
     /// The captured value, `nil` for no-capture
     public var value: Any? {
-      // FIXME: Should this return the substring for default-typed
-      // values?
-      representation.value
+      representation.value ?? substring
+    }
+
+    internal var type: Any.Type {
+      value.map { Swift.type(of: $0) } ?? Substring.self
```
Comment (milseman) [inline Sources/_StringProcessing/Regex/AnyRegexOutput.swift:69]: What about optional Substring?

File: Sources/_StringProcessing/Regex/Match.swift [hunk 4]
```diff
@@ -37,33 +41,48 @@ extension Regex.Match {
...
       let typeErasedMatch = anyRegexOutput.existentialOutput(
-        from: anyRegexOutput.input[range]
+        from: input[range]
       )
       return typeErasedMatch as! Output
     }
   }
 
+  var wholeMatchType: Any.Type {
+    value.map { type(of: $0) } ?? Substring.self
```
Comment (milseman) [inline Sources/_StringProcessing/Regex/Match.swift:69]: What if it's optional? Also, are we tracking nested optionals correctly here (I.e. do they get turned into values or are they counted)?

**Review by milseman — COMMENTED** [review]
It might be useful to write tests where a capture transform produces a `Substring` by e.g. returning `.dropFirst()`. Similarly for when that capture-transform appears inside an `Optionally` or is adding or removing a layer of Substring optionality.

File: Sources/_StringProcessing/Regex/AnyRegexOutput.swift (hunk 3 above)
Comment (rxwei) [reply Sources/_StringProcessing/Regex/AnyRegexOutput.swift]: Good point. Fixed.

File: Sources/_StringProcessing/Regex/Match.swift (hunk 4 above)
Comment (rxwei) [reply Sources/_StringProcessing/Regex/Match.swift:69]: When there's no dedicated value for whole match, it can only be a `Substring`.

**Pull request merged**